### ISSUE TITLE
[data grid][question] Customize 'Select All' component and behavior in the header

### ISSUE THREAD

**Issue by igor-serzhan** [mui/mui-x#3685]
### Summary 💡

Currently it's not possible to specify component or custom callback for the 'Select All' checkbox in the header column of Data Grid. 
`GridHeaderCheckbox` uses `BaseCheckbox` slot and incapsulated logic for the `handleChange` callback at the current implementation.

'Select All' checkbox selects only the top level rows in the Data Grid with groups while the combination of 'Ctrl' + 'A' keys select all rows (parents and childrens)

### Examples 🌈

It would be great to pass `HeaderCheckbox` component or at least `onSelectAllChange` callback to the DataGrid component.

Here is the example of the issue with selecting childrens
![Screen Recording 2022-01-21 at 11 52 58](https://user-images.githubusercontent.com/95248354/150463451-ce081b21-571f-4a4e-b454-83d8af69c552.gif)

### Motivation 🔦

First of all, with current implementation it's not possible to select all collapsed childrens at the Tree Data DataGrid (only with `Ctrl + A` keyboard combination).

Secondly, to render custom header for the `Select All` column you need to find header checkbox by `gridClasses.columnHeaderCheckbox`, use refs and appendChild DOM API which looks quite complicated.

Here is the example of custom header.

<img width="644" alt="Screenshot 2022-01-21 at 11 49 53" src="https://user-images.githubusercontent.com/95248354/150462501-214d2f27-f84c-452e-8692-90e3611fad60.png">

### Order ID 💳 (optional)

35653

**Comment by alexfauquette** [comment]
Thanks for this issue. I agree being able to customize the select all checkbox is a nice to have, especially if you want to make complex ones, such as the G-mail select all
![image](https://user-images.githubusercontent.com/45398769/150494258-bce71672-84cf-406d-8449-b58c4e474e4f.png)

**Comment by DanailH** [comment]
@igor-serzhan thanks for reporting this! I agree that is a nice feature that can be added, but in the meantime, you can probably achieve this by yourself. You can provide your own checkbox component to the `checkboxSelection` column type via the `renderHeader` property and use the `selectionModel` and the `onSelectionModelChange` props to control the selection yourself. 

You can check this example to see how to do that -> https://codesandbox.io/s/renderheadergrid-material-demo-forked-vmkcf

More info regarding the `renderHeader` can be found here -> https://mui.com/components/data-grid/columns/#render-header

More info regarding the `selectionModel` and the `onSelectionModelChange` can be found here -> https://mui.com/components/data-grid/selection/#controlled-selection

**Comment by jaballogian** [comment]
> ### Duplicates
> * [x]  I have searched the existing issues
> 
> ### Latest version
> * [x]  I have tested the latest version
> 
> ### Summary 💡
> Currently it's not possible to specify component or custom callback for the 'Select All' checkbox in the header column of Data Grid. `GridHeaderCheckbox` uses `BaseCheckbox` slot and incapsulated logic for the `handleChange` callback at the current implementation.
> 
> 'Select All' checkbox selects only the top level rows in the Data Grid with groups while the combination of 'Ctrl' + 'A' keys select all rows (parents and childrens)
> 
> ### Examples 🌈
> It would be great to pass `HeaderCheckbox` component or at least `onSelectAllChange` callback to the DataGrid component.
> 
> Here is the example of the issue with selecting childrens ![Screen Recording 2022-01-21 at 11 52 58](https://user-images.githubusercontent.com/95248354/150463451-ce081b21-571f-4a4e-b454-83d8af69c552.gif)
> 
> ### Motivation 🔦
> First of all, with current implementation it's not possible to select all collapsed childrens at the Tree Data DataGrid (only with `Ctrl + A` keyboard combination).
> 
> Secondly, to render custom header for the `Select All` column you need to find header checkbox by `gridClasses.columnHeaderCheckbox`, use refs and appendChild DOM API which looks quite complicated.
> 
> Here is the example of custom header.
> 
> <img alt="Screenshot 2022-01-21 at 11 49 53" width="644" src="https://user-images.githubusercontent.com/95248354/150462501-214d2f27-f84c-452e-8692-90e3611fad60.png">
> 
> ### Order ID 💳 (optional)
> 35653

I also agree that there would be nice if the user selects/checks one of the parent checkboxes (for example "Helena Ryan" from the example above), all of its children (196 children) are also selected. If the user unselects one of the children from its parent (for example unselects "Trevor Perkins" from its parent "Helena Ryan"), its parent is also unselected.

I'm creating this feature manually using the `checkboxSelection` and `onSelectionModelChange` props but haven't succeeded yet. Please let me know the code if one of you has succeeded in implementing it.

**Comment by DanailH** [comment]
@igor-serzhan did you check my answer?

**Comment by kvenkatasivareddy** [comment]
Any update on this @DanailH @jaballogian @igor-serzhan ?

**Comment by EdmundsEcho** [comment]
The improvements on the grid, and the grid itself is really, really good.  That said, I'm surprised that with all of the updated versions this feature has not been added.  I just want to know when the user wants to selectAll or, deselectAll.  Relying on the selectionModel and its changes is missing an important use case we have described here: when not all of the rows are displayed.  Just being able to "catch" that selectAll event is all I need to get stuff done.  Until then, I'll continue to resort to my custom render header...

**Comment by ewarrenG** [comment]
> @igor-serzhan thanks for reporting this! I agree that is a nice feature that can be added, but in the meantime, you can probably achieve this by yourself. You can provide your own checkbox component to the `checkboxSelection` column type via the `renderHeader` property and use the `selectionModel` and the `onSelectionModelChange` props to control the selection yourself.
> 
> You can check this example to see how to do that -> https://codesandbox.io/s/renderheadergrid-material-demo-forked-vmkcf
> 
> More info regarding the `renderHeader` can be found here -> https://mui.com/components/data-grid/columns/#render-header
> 
> More info regarding the `selectionModel` and the `onSelectionModelChange` can be found here -> https://mui.com/components/data-grid/selection/#controlled-selection

The sandbox link is broken

**Comment by michelengelen** [comment]
Thanks for the report, and thanks to everyone for the patience. All parts of this request are now supported:
- Custom "Select All" header: pass your own checkbox column with `GRID_CHECKBOX_SELECTION_COL_DEF` and `renderHeader`. See the custom checkbox column docs.
- Select-all callback: subscribe to the `headerSelectionCheckboxChange` event through `apiRef`.
- Selecting children in tree data: the `rowSelectionPropagation` prop shipped in v7.19.0. It is enabled by default since v8. "Select All" and parent checkboxes now select collapsed children too.

I will close this issue. Please open a new issue if your use case is still not covered.
